File: src/main/java/task7_2_HD_s224384754/TodoService.java
package task7_2_HD_s224384754;

import java.util.*;
import java.util.concurrent.atomic.AtomicLong;

public class TodoService {
    private final Map<Long, Todo> todos = new HashMap<>();
    private final AtomicLong idGenerator = new AtomicLong(1);

    public Todo create(String title) {
        if (title == null || title.trim().isEmpty()) {
            throw new IllegalArgumentException("Title cannot be empty");
        }
        long id = idGenerator.getAndIncrement();
        Todo todo = new Todo(id, title, false);
        todos.put(id, todo);
        return todo;
    }

    public List<Todo> getAll() {
        return new ArrayList<>(todos.values());
    }

    public Todo getById(long id) {
        Todo todo = todos.get(id);
        if (todo == null) {
            throw new NoSuchElementException("Todo not found: " + id);
        }
        return todo;
    }

    public Todo complete(long id) {
        Todo todo = getById(id);
        todo.setCompleted(true);
        return todo;
    }

    public boolean delete(long id) {
        return todos.remove(id) != null;
    }

    public int count() {
        return todos.size();
    }
}

File: todo-app-python/todo_service.py
"""Business logic for managing todos."""


class TodoService:
    """In-memory CRUD service for Todo items."""

    def __init__(self):
        # All todos stored as {id: dict}
        self._todos = {}
        self._next_id = 1

    def create(self, title):
        """Create a new todo with the given title."""
        if title is None or title.strip() == "":
            raise ValueError("Title cannot be empty")
        todo = {
            "id": self._next_id,
            "title": title,
            "completed": False,
        }
        self._todos[self._next_id] = todo
        self._next_id += 1
        return todo

    def get_all(self):
        """Return all todos as a list."""
        return list(self._todos.values())

    def get_by_id(self, todo_id):
        """Return a todo by its id, or raise KeyError if not found."""
        if todo_id not in self._todos:
            raise KeyError(f"Todo not found: {todo_id}")
        return self._todos[todo_id]

    def complete(self, todo_id):
        """Mark a todo as completed."""
        todo = self.get_by_id(todo_id)
        todo["completed"] = True
        return todo

    def delete(self, todo_id):
        """Delete a todo by id, return True if deleted."""
        if todo_id in self._todos:
            del self._todos[todo_id]
            return True
        return False

    def count(self):
        """Return the number of stored todos."""
        return len(self._todos)